src/mugs.js

```diff
@@ -78,8 +78,116 @@ define([
         getIcon: function () {
             return this.options.getIcon(this);
         },
+        /**
+         * Validate mug
+         *
+         * This method may fire a "messages-changed" event.
+         *
+         * @param attr - The property to validate. All properties will
+         *      be validated if this argument is omitted.
+         * @returns - True if validation messages changed else false.
+         */
+        validate: function (attr) {
+            var mug = this,
+                changed;
+            this.form.updateLogicReferences(this, attr);
+            if (attr) {
+                changed = this._validate(attr);
+            } else {
+                _.each(_.keys(this.p.__data), function (attr) {
+                    changed = changed || mug._validate(attr);
+                });
+            }
+            if (changed) {
+                this.fire({type: "messages-changed", mug: this});
+            }
+            return changed;
+        },
+        _validate: function (attr) {
+            var mug = this,
+                spec = mug.spec[attr];
+            if (!spec) {
+                // should throw error?
+                window.console.log("unexpected property: " + attr);
+                return false;
+            }
+            var value = mug.p[attr],
+                label = spec.lstring || attr,
+                message = "";
+
+            // TODO use data.hasOwnProperty(attr) rather than !value?
+            if (!value && spec.presence === 'required') {
+                // can the user always fix this error?
+                message = label + ' is required.';
+            } else if (value && spec.presence === 'notallowed') {
+                // can the user always fix this error?
+                message = label + ' is not allowed.';
+            } else if (spec.validationFunc) {
+                try {
+                    message = spec.validationFunc(mug);
+                } catch (err) {
+                    // this should never happen
+                    message = label + " validation failed\n" + util.formatExc(err);
+                }
+                if (message === "pass") {
```

Comment:
Returning `"pass"` is an artifact of the past. I agree it would be nicer to simply return null, but don't want to change all of the places that depend on it in this PR.